cmd_migrate_db.go

```diff
@@ -0,0 +1,707 @@
+			dbName = lncfg.SqliteChainDBName
+
+		case lncfg.NSMacaroonDB:
+			path = walletDir
+			dbName = lncfg.SqliteChainDBName
+
+		case lncfg.NSDecayedLogDB:
+			path = graphDir
+			dbName = lncfg.SqliteChannelDBName
+
+		case lncfg.NSTowerClientDB:
+			path = graphDir
+			dbName = lncfg.SqliteChannelDBName
+
+		case lncfg.NSTowerServerDB:
+			path = towerServerDir
+			dbName = lncfg.SqliteChannelDBName
+
+		case lncfg.NSNeutrinoDB:
+			path = walletDir
+			dbName = lncfg.SqliteNeutrinoDBName
+		}
+
+		args = []interface{}{
+			context.Background(),
+			&sqlite.Config{
+				Timeout: time.Minute,
+			},
+			path,
+			dbName,
+			prefix,
+		}
+
+		logger.Infof("Opening sqlite backend at %s  "+
+			"for prefix '%s'", filepath.Join(path, dbName),
+			prefix)
+
+	default:
+		return nil, fmt.Errorf("unknown backend: %v", backend)
+	}
+
+	return kvdb.Open(backend, args...)
+}
+
+// checkMarkerPresent checks if a marker is present in the database.
+func checkMarkerPresent(db walletdb.DB, markerKey []byte) ([]byte, error) {
+	rtx, err := db.BeginReadTx()
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = rtx.Rollback() }()
+
+	return channeldb.CheckMarkerPresent(rtx, markerKey)
+}
+
+// addMarker adds a marker to the database.
+func addMarker(db walletdb.DB, markerKey []byte) error {
+	rwtx, err := db.BeginReadWriteTx()
+	if err != nil {
+		logger.Errorf("Failed to begin read write transaction: %v", err)
+		return err
+	}
+
+	markerValue := []byte(fmt.Sprintf("lndinit migrate-db %s", time.Now()))
+	if err := channeldb.AddMarker(rwtx, markerKey, markerValue); err != nil {
+		return err
+	}
+
+	return rwtx.Commit()
+}
+
+// createWalletMarker creates a marker in the wallet database to indicate it's
+// ready for use. This is only needed for non-bolt databases.
+func createWalletMarker(db walletdb.DB, logger btclog.Logger) error {
+	logger.Info("Creating 'wallet created' marker")
+
+	tx, err := db.BeginReadWriteTx()
+	if err != nil {
+		return fmt.Errorf("failed to begin transaction: %w", err)
+	}
+	defer func() { _ = tx.Rollback() }()
+
+	metaBucket, err := tx.CreateTopLevelBucket([]byte(walletMetaBucket))
+	if err != nil {
+		return fmt.Errorf("failed to create meta bucket: %w", err)
+	}
+
+	err = metaBucket.Put([]byte(walletReadyKey), []byte(walletReadyKey))
+	if err != nil {
+		return fmt.Errorf("failed to put wallet ready marker: %w", err)
+	}
+
+	logger.Info("Committing 'wallet created' marker")
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("error committing 'wallet created' "+
+			"marker: %w", err)
+	}
+
+	return nil
+}
+
+// checkChannelDBMigrationsApplied checks if the channel DB migrations are
+// applied.
+func checkChannelDBMigrationsApplied(db walletdb.DB) error {
+	var meta channeldb.Meta
+	err := kvdb.View(db, func(tx kvdb.RTx) error {
+		return channeldb.FetchMeta(&meta, tx)
+	}, func() {
+		meta = channeldb.Meta{}
+	})
+	if err != nil {
+		return err
+	}
+
+	if meta.DbVersionNumber != channeldb.LatestDBVersion() {
```

Comment:
I just wanted to do one last circle back to think about DB version numbers to make sure the tool handles future migrations well.

Is this check good enough to handle if LND is upgraded but lndinit is an older version? Seems like it is already good in the case where lndinit is upgraded but LND is an older version.

I guess generally you will want to make clear that this tool needs to stay up to date with LND, and that you should update LND first before running a DB migration.

Reply:
The idea is that we'll integrate this code into `lnd` with 0.20. To have it externally in `lndinit` during the 0.19 period is mostly to not force anyone that doesn't want to migrate yet into doing so. But give an option to all those who want to. Then, assuming we get a lot of testers running this voluntarily with success, we also have a lot of confidence for doing it automatically with 0.20. So I don't think we'll need to care about compatibility between this tool and `lnd`.